src/libostree/ostree-fetcher-curl.c

```diff
@@ -575,6 +591,178 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+      "Sun",
+    };
+  size_t day_name_index;
+  const char *month_names[] =
+    {
+      "Jan",
+      "Feb",
+      "Mar",
+      "Apr",
+      "May",
+      "Jun",
+      "Jul",
+      "Aug",
+      "Sep",
+      "Oct",
+      "Nov",
+      "Dec",
+    };
+  size_t month_name_index;
+
+  if (len != 29)
+    return NULL;
+
+  const char *day_name = buf;
+  const char *day = buf + 5;
+  const char *month_name = day + 3;
+  const char *year = month_name + 4;
+  const char *hour = year + 5;
+  const char *minute = hour + 3;
+  const char *second = minute + 3;
+  const char *tz = second + 3;
+
+  for (day_name_index = 0; day_name_index < G_N_ELEMENTS (day_names); day_name_index++)
+    {
+      if (strncmp (day_names[day_name_index], day_name, 3) == 0)
+        break;
+    }
+  if (day_name_index >= G_N_ELEMENTS (day_names))
+    return NULL;
+  /* don’t validate whether the day_name matches the rest of the date */
+  if (*(day_name + 3) != ',' || *(day_name + 4) != ' ')
+    return NULL;
+  if (!parse_uint (day, 2, 1, 31, &day_int))
+    return NULL;
+  if (*(day + 2) != ' ')
+    return NULL;
+  for (month_name_index = 0; month_name_index < G_N_ELEMENTS (month_names); month_name_index++)
+    {
+      if (strncmp (month_names[month_name_index], month_name, 3) == 0)
+        break;
+    }
+  if (month_name_index >= G_N_ELEMENTS (month_names))
+    return NULL;
+  if (*(month_name + 3) != ' ')
+    return NULL;
+  if (!parse_uint (year, 4, 0, 9999, &year_int))
+    return NULL;
+  if (*(year + 4) != ' ')
+    return NULL;
+  if (!parse_uint (hour, 2, 0, 23, &hour_int))
+    return NULL;
+  if (*(hour + 2) != ':')
+    return NULL;
+  if (!parse_uint (minute, 2, 0, 59, &minute_int))
+    return NULL;
+  if (*(minute + 2) != ':')
+    return NULL;
+  if (!parse_uint (second, 2, 0, 60, &second_int))  /* allow leap seconds */
+    return NULL;
+  if (*(second + 2) != ' ')
+    return NULL;
+  if (strncmp (tz, "GMT", 3) != 0)
+    return NULL;
+
+  return g_date_time_new_utc (year_int, month_name_index + 1, day_int,
+                              hour_int, minute_int, second_int);
+}
+
+/* CURLOPT_HEADERFUNCTION */
+static size_t
+response_header_cb (const char *buffer, size_t size, size_t n_items, void *user_data)
+{
+  const size_t real_size = size * n_items;
+  GTask *task = G_TASK (user_data);
+  FetcherRequest *req;
+
+  req = g_task_get_task_data (task);
+
+  if (real_size > strlen ("ETag: ") &&
+      strncasecmp (buffer, "ETag: ", strlen ("ETag: ")) == 0)
+    {
+      g_clear_pointer (&req->out_etag, g_free);
+      req->out_etag = g_strstrip (g_strdup (buffer + strlen ("ETag: ")));
+    }
+  else if (real_size > strlen ("Last-Modified: ") &&
+           strncasecmp (buffer, "Last-Modified: ", strlen ("Last-Modified: ")) == 0)
+    {
+      g_autoptr(GDateTime) dt = NULL;
+      const char *lm_buf = buffer + strlen ("Last-Modified: ");
+      size_t lm_len = real_size - strlen ("Last-Modified: ");
+
+      /* Check termination */
```

Comment:
Why do we do this in the last-modified case but not etag?

Reply:
Simplicity of the code; stripping the termination is handled in the ETag case by `g_strstrip()`, and it seemed overkill to check that the last two bytes actually were `\r\n`, given that the set of allowed characters in an ETag is basically everything, and that all we do with it is to store it in an xattr and then regurgitate it. My thinking was that as long as we’re not actually parsing the ETag (and correctly handle it as an arbitrary byte array) it doesn’t need validating, just storing and regurgitating.

I can tighten up the parsing on the ETag if you want, though; I think the HTTP spec is not 100% liberal with it.

Reply:
Looking at https://stackoverflow.com/a/50180173, it seems like it's possible to have arbitrary amounts of whitespace before the value as well. And that'll throw off the `len != 29` check in the parser. So maybe simpler to just `g_strstrip (g_strdup (...))` here too.

Reply:
Done